### PR TITLE
Handle errors when inferring separator of csv

### LINKED ISSUE #44060: Comma in quoted string during CSV upload: CSV error (unexpected character: ,)
### Describe the bug

Trying to upload the following file will result in an error:
```csv
"c1","c2"
"a,b,c","d"
```

```
CSV error (unexpected character: ,)
```

### To Reproduce

1. Run Metabase 1.50.0
2. Go to http://localhost:3000/admin/settings/uploads
3. Select / enter "Sample Database", "PUBLIC" and "upload_"
4. Go to http://localhost:3000/collection/root
5. Click on upload icon in the top right ("Upload data to Our analytics")
6. Select the file quoted above
7. See error

### Expected behavior

File should be processed, since comma in string is quoted.

### Logs

```
2024-06-12 13:57:08,992 ERROR middleware.log :: POST /api/card/from-csv 500 164.9 ms (5 DB calls) {:metabase-user-id 1}
{:message "CSV error (unexpected character: ,)"}
```

### Information about your Metabase installation

```JSON
{
  "browser-info": {
    "language": "en-US",
    "platform": "Linux x86_64",
    "userAgent": "Mozilla/5.0 (X11; Linux x86_64; rv:126.0) Gecko/20100101 Firefox/126.0",
    "vendor": ""
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "11.0.23+9",
    "java.vendor": "Eclipse Adoptium",
    "java.vendor.url": "https://adoptium.net/",
    "java.version": "11.0.23",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "11.0.23+9",
    "os.name": "Linux",
    "os.version": "6.8.11-300.fc40.x86_64",
    "user.language": "en",
    "user.timezone": "GMT"
  },
  "metabase-info": {
    "databases": [
      "h2",
      "postgres"
    ],
    "hosting-env": "unknown",
    "application-database": "postgres",
    "application-database-details": {
      "database": {
        "name": "PostgreSQL",
        "version": "14.10"
      },
      "jdbc-driver": {
        "name": "PostgreSQL JDBC Driver",
        "version": "42.7.3"
      }
    },
    "run-mode": "prod",
    "version": {
      "date": "2024-06-10",
      "tag": "v1.50.0",
      "hash": "4fe42d3"
    },
    "settings": {
      "report-timezone": null
    }
  }
}
```

### Severity

blocking some users

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by dpsutton**
Fixes https://github.com/metabase/metabase/issues/44060

This file

```
"c1","c2"
"a,b,c","d"
```

was failing to upload for a silly reason: when we were checking for which separator it used, we had a parse error when using a semi-colon and that blew up the whole pipeline.

### CODE REVIEW

**Review by johnswanson — APPROVED** [review]
I didn't actually know about this arity of `into` that takes a transducer, so it was a little less clear to me at first - but I think it's on me to get more comfortable with transducers! LGTM.

File: test/metabase/upload_test.clj [hunk 1]
```diff
@@ -457,7 +457,45 @@
...
+            (is (=? (cond->> [["c1" {:effective_type :type/Text}]
+                              ["c2" {:effective_type :type/Text}]]
+                      (driver.u/supports? driver/*driver* :upload-with-auto-pk (mt/db))
+                      (cons ["_mb_row_id" {:semantic_type     :type/PK
+                                           :base_type         :type/BigInteger}]))
+                    (->> (t2/select :model/Field :table_id (:id table))
+                         (sort-by :database_position)
+                         (map (juxt (comp u/lower-case-en :name) identity))))))
+          (testing "Check the data was uploaded into the table"
+            (is (= 1 (count (rows-for-table table))))))))))
```
Comment (calherries) [inline test/metabase/upload_test.clj:479]: Small thing: the test that you copied from `create-from-csv-test` tests more than we need to test here. We just need to check the contents of the uploaded table is what we expected. We can do that like this with less code:

```suggestion
(deftest infer-separator-catch-exception-test
  (testing "when trying one of the separators in [[upload/infer-separator]] fails the upload should still work end to end (#44034)"
    (mt/test-drivers (mt/normal-drivers-with-feature :uploads)
      (with-mysql-local-infile-on-and-off
        (with-upload-table!
          [table (create-from-csv-and-sync-with-defaults!
                  :file (csv-file-with ["\"c1\",\"c2\""
                                        "\"a,b,c\",\"d\""]))]
          (testing "Check the data was uploaded into the table correctly"
            (is (= (header-with-auto-pk ["c1", "c2"])
                   (column-names-for-table table)))
            (is (= (rows-with-auto-pk [["a,b,c" "d"]])
                   (rows-for-table table)))))))))
```

**Review by calherries — APPROVED** [review]
LGTM, thanks for picking this up!

File: test/metabase/upload_test.clj (hunk 1 above)
Comment (dpsutton) [reply test/metabase/upload_test.clj]: super helpful.

**Pull request merged**